HTTP AutoClient NuGet package content (#36970)
dotnet/docs: docs/fundamentals/networking/http/http-autoclient.md

## Auto client's dependency injection hooks

Along with the interface's implementation, extension methods are generated to register it into the dependency injection container. The name of the generated extension method is the same as your interface name, replacing the leading `I` with `Add`.

For example, consider the following interface definition:

```csharp
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.Extensions.Http.AutoClient;

[AutoClient(nameof(IUserClient))]
public interface IUserClient
{
    [Get("/api/users")]
    public Task<List<User>> GetUsersAsync(
        CancellationToken cancellationToken = default);
}
```

You'd have registration code similar to the following:

```csharp
var builder = Host.CreateApplicationBuilder(args);

builder.Services
    .AddHttpClient(
        nameof(IUserClient),
        client => client.BaseAddress = new Uri("<The REST API base address>"))
    .AddHttpClientMetering()
    .AddHttpClientLogging()
    .AddStandardResilienceHandler()
    .AddSocketsHttpHandler();

services.AddUserClient();
```

Then services would consume the client through its constructor:

```csharp

> public sealed class UserService
> {
>     private readonly IUserClient _userClient;
> 
>     public MyService(IUserClient userClient) => _userClient = userClient;
> 
>     public async Task ProcessUsersAsync()
>     {
>         var users = await _myClient.GetUsersAsync();
>         // ...
>     }
> }

Reviewer: Suggestion for a primary constructor:

```suggestion
public sealed class UserService(IUserClient userClient)
{
    public async Task ProcessUsersAsync()
    {
        var users = await userClient.GetUsersAsync();
        // ...
    }
}
```

```csharp
var builder = Host.CreateApplicationBuilder(args);

// Add the named HttpClient to the service collection
builder.Services.AddHttpClient(
    nameof(IUserClient),
    client =>
    {
        client.BaseAddress = new Uri("<The REST API base address>");
    });

// Add the IUserClient/UserClient to the service collection
builder.Services.AddUserClient();
```

In the preceding example code:

- The <xref:Microsoft.Extensions.Hosting.Host.CreateApplicationBuilder(System.String[])?displayProperty=fullName> is used to create a <xref:Microsoft.Extensions.Hosting.HostApplicationBuilder>.
- The <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection> is retrieved from the <xref:Microsoft.Extensions.Hosting.HostApplicationBuilder.Services?displayProperty=nameWithType> property, to call the `AddHttpClient` extension method.
  - The `AddHttpClient` extension method is called with the name of the <xref:System.Net.Http.HttpClient> to be registered, and a delegate to configure the <xref:System.Net.Http.HttpClient> instance.
- The `AddUserClient` extension method is called to register the <xref:IUserClient> interface and its implementation.

> Consuming the client is as simple as injecting it into your service's constructor:

Reviewer: nit:

```suggestion
You can consume the client by injecting it into your service's constructor:
```
